File: src/main/java/net/ins/edu/algorithms/j/search/BinarySearch.java
package net.ins.edu.algorithms.j.search;

public class BinarySearch<T extends Comparable<T>> {

    public BinarySearch(T[] arr) {
        this.arr = arr;
    }

    private T[] arr;

    public int find(T search) {
        return find(0, arr.length - 1, search);
    }

    private int find(int low, int high, T search) {
        if (low > high) {
            return -1;
        }
        int mid = low + (high - low) / 2;
        if (arr[mid].equals(search)) {
            return mid;
        }
        if (search.compareTo(arr[mid]) > 0) {
            return find(mid + 1, high, search);
        } else {
            return find(low, mid + 1, search);
        }
    }


    public static void main(String[] args) {
        Integer[] arr = new Integer[] {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
        BinarySearch<Integer> search = new BinarySearch<>(arr);
        System.out.println(search.find(8));;
    }
}

File: python/search/binary_search.py

from array import array


def binary_search(arr: array, val: int, low: int, high: int):
    if low > high:
        return -1
    
    mid = int(low + (high-low)/2)

    if arr[mid] == val:
        return mid

    return binary_search(arr, val, low, mid-1) if val < arr[mid] else binary_search(arr, val, mid+1, high)

if __name__ == '__main__':
    print(binary_search([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 5, 0, 9))
    print(binary_search([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 1, 0, 9))
    print(binary_search([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 10, 0, 9))
    print(binary_search([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 11, 0, 9))